dags/trending-projects-snapshot-dag.py

```diff
@@ -0,0 +1,183 @@
+def trending_projects_snapshot() -> None:
+    """
+    This DAG executes a query on Snowflake to retrieve information about trending public projects and 
+    reports the results to a Synapse table.
+
+    The main steps performed in this DAG are:
+    1. Query Snowflake to identify public projects, their file sizes, and recent download activities.
+    2. Aggregate the results to determine num. of unique users, last download date, and total data size.
+    3. Format the results into a list of SnapshotsMetric objects.
+
+    The Snowflake connection and Synapse connection IDs, as well as the time window for recent downloads,
+    are provided via DAG parameters.
+    """
+
+    @task
+    def get_trending_project_snapshot(**context) -> List[SnapshotMetrics]:
+        """Execute the query on Snowflake and return the results."""
+        snow_hook = SnowflakeHook(context["params"]["snowflake_conn_id"])
+        ctx = snow_hook.get_conn()
+        cs = ctx.cursor()
+        query = f"""
+                WITH PUBLIC_PROJECTS AS (
+                    SELECT PROJECT_ID
+                    FROM SYNAPSE_DATA_WAREHOUSE.SYNAPSE.NODE_LATEST
+                    WHERE 1=1
+                    AND NODE_TYPE = 'project'
+                    AND IS_PUBLIC = TRUE
+                ),
+                LATEST_FILE_HANDLES AS (
+                    SELECT DISTINCT ID, PROJECT_ID, FILE_HANDLE_ID
+                    FROM SYNAPSE_DATA_WAREHOUSE.SYNAPSE.NODE_LATEST
+                    WHERE 1=1
+                    AND NODE_TYPE = 'file'
+                    AND PROJECT_ID IN (SELECT PROJECT_ID FROM PUBLIC_PROJECTS)
+                ),
+                FILE_SIZES AS (
+                    SELECT fh.PROJECT_ID, SUM(f.CONTENT_SIZE) / POWER(1024, 3) AS TOTAL_DATA_SIZE_IN_GIB
+                    FROM LATEST_FILE_HANDLES fh
+                    JOIN SYNAPSE_DATA_WAREHOUSE.SYNAPSE.FILE_LATEST f
+                    ON fh.FILE_HANDLE_ID = f.ID
+                    GROUP BY fh.PROJECT_ID
+                ),
+                RECENT_DOWNLOADS AS (
+                    SELECT PROJECT_ID, RECORD_DATE, USER_ID
+                    FROM SYNAPSE_DATA_WAREHOUSE.SYNAPSE.FILEDOWNLOAD
+                    WHERE 1=1
+                    AND RECORD_DATE > DATEADD(DAY, -{context["params"]["time_window"]}, CURRENT_DATE()) 
+                    AND STACK = 'prod'
+                    AND PROJECT_ID IN (SELECT PROJECT_ID FROM PUBLIC_PROJECTS)
+                )
+
+                SELECT 
+                    pp.PROJECT_ID,
+                    COUNT(DISTINCT rd.USER_ID) AS N_UNIQUE_USERS,
+                    COALESCE(TO_CHAR(MAX(rd.RECORD_DATE), 'YYYY-MM-DD'), 'N/A') AS LAST_DOWNLOAD_DATE,
+                    ROUND(fs.TOTAL_DATA_SIZE_IN_GIB, 3) AS TOTAL_DATA_SIZE_IN_GIB
+                FROM 
+                    PUBLIC_PROJECTS pp
+
+                -- Join RECENT_DOWNLOADS
+                LEFT JOIN RECENT_DOWNLOADS rd
+                ON pp.PROJECT_ID = rd.PROJECT_ID
```

Comment:
Nit: this is just personal preference, but short hand is never easier to read than just writing the whole table out:


```
FROM 
    PUBLIC_PROJECTS
-- Join RECENT_DOWNLOADS
LEFT JOIN RECENT_DOWNLOADS
ON PUBLIC_PROJECTS.PROJECT_ID = RECENT_DOWNLOADS.PROJECT_ID
```